presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushPartialAggregationThroughExchange.java

```diff
@@ -313,20 +316,40 @@ private PlanNode split(AggregationNode node, Context context)
                 ImmutableList.of(),
                 FINAL,
                 node.getHashVariable(),
-                node.getGroupIdVariable());
+                node.getGroupIdVariable(),
+                Optional.of(aggregationId));
     }
 
-    private boolean partialAggregationNotUseful(AggregationNode aggregationNode, ExchangeNode exchangeNode, Context context)
+    private boolean partialAggregationNotUseful(AggregationNode aggregationNode, ExchangeNode exchangeNode, Context context, int numAggregationKeys)
     {
         StatsProvider stats = context.getStatsProvider();
         PlanNodeStatsEstimate exchangeStats = stats.getStats(exchangeNode);
         PlanNodeStatsEstimate aggregationStats = stats.getStats(aggregationNode);
-        double inputBytes = exchangeStats.getOutputSizeInBytes(exchangeNode);
-        double outputBytes = aggregationStats.getOutputSizeInBytes(aggregationNode);
+        double inputSize = exchangeStats.getOutputSizeInBytes(exchangeNode);
+        double outputSize = aggregationStats.getOutputSizeInBytes(aggregationNode);
+        PartialAggregationStatsEstimate partialAggregationStatsEstimate = aggregationStats.getPartialAggregationStatsEstimate();
+        boolean isConfident = exchangeStats.isConfident();
+        // keep old behavior of skipping partial aggregation only for single-key aggregations
+        boolean numberOfKeyCheck = usePartialAggregationHistory(context.getSession()) || numAggregationKeys == 1;
+        if (partialAggregationStatsEstimate != PartialAggregationStatsEstimate.unknown() && usePartialAggregationHistory(context.getSession())) {
```

Comment:
nit: maybe add a helper method `isUnknown` to avoid comparing the references (it can be potentially fragile)